lib/Runtime/Library/JSON.cpp

```diff
@@ -682,56 +681,90 @@ namespace JSON
                 Js::JavascriptStaticEnumerator enumerator;
                 if (object->GetEnumerator(&enumerator, EnumeratorFlags::SnapShotSemantics | EnumeratorFlags::EphemeralReference, scriptContext))
                 {
-                    Js::DynamicObject * dynamicObject = ((Js::DynamicObject*)object);
-                    if (ReplacerFunction != replacerType || !dynamicObject->HasObjectArray())
+                    bool isDynamicObject = Js::DynamicObject::Is(object);
+                    Js::DynamicObject * dynamicObject = nullptr;
+                    bool hasObjectArray = false;
+                    if (isDynamicObject)
                     {
-                        uint32 propertyCount = dynamicObject->GetPropertyCount();
+                        dynamicObject = Js::DynamicObject::FromVar(object);
+                        if (dynamicObject->HasObjectArray())
+                        {
+                            hasObjectArray = dynamicObject->HasObjectArray();
+                        }
+                    }
 
-                        result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
-                        Js::DynamicTypeHandler * typeHandler = dynamicObject->GetTypeHandler();
+                    if (ReplacerFunction != replacerType || !hasObjectArray)
+                    {
+                        uint32 propertyCount;
 
-                        // if object has an objectarray, (aka indexed properties)
+                        // if object has an objectArray, (aka indexed properties)
                         // we need to loop them first since they won't show up on the second loop below
-                        if (dynamicObject->HasObjectArray())
+                        // besides, we might be talking to a non dynamicObject
+                        // in that case, we will also take the enumerator path
+                        if (!isDynamicObject || hasObjectArray)
                         {
+                            // get the precise count because resizing ConcatString will be expensive!
+                            propertyCount = (uint32) this->GetPropertyCount(object, &enumerator);
+                            result = Js::ConcatStringBuilder::New(this->scriptContext, propertyCount);
+
                             enumerator.Reset();
-                            while ((propertyName = enumerator.MoveAndGetNext(id)) != NULL)
+                            while (propertyCount > 0 && (propertyName = enumerator.MoveAndGetNext(id)) != NULL)
                             {
                                 // if unsuccessful get propertyId from the string
                                 scriptContext->GetOrAddPropertyRecord(propertyName, &propRecord);
-                                if (!propRecord->IsNumeric()) break;
+                                if (isDynamicObject && !propRecord->IsNumeric())
```

Comment:
What happens if a non-`DynamicObject` has non-numeric properties here? Why do we treat that differently to a `DynamicObject` with non-numeric properties?

Reply:
when it's non-dynamicObject, we want to use this enumator loop only because overridden type might behave differently.